### PR TITLE
Add nrf5 platform

### PR DESCRIPTION
**Pull request by d00616**
This is the support of the Nordic Semiconductor NRF51 (nrf51822, nrf51422) and NRF52 (nrf52832) platform. Based on arduino-nRF5(https://github.com/sandeepmistry/arduino-nRF5).

The Radio is working in an NRF24 compatible mode. Other protocols can implemented. SoftDevice (BLE) is not supported and must be removed before flashing MySensors.
 
The included "NVM" driver allows an EEPROM like access to the last 16/32k of Flash pages. It can easily ported to ESP8266 or SAMD by providing an Flash interface.

Documentation about using this code: https://www.openhardware.io/view/376/MySensors-NRF5-Platform

### CODE REVIEW

**Comment by henrikekblad** [comment]
Great job @d00616!

**Comment by d00616** [comment]
> Great job @d00616 [1]!

Thank you. It was a long way for me to implement an open source 
compatible version without using Nordic's SDK with it's own license.

**Comment by fallberg** [comment]
Nice work! Once submitted I'll at some point update jenkins to do some sanity checking on this variant so we get static code analysis and doxygen validation on it fully covered. I might make various patches to the code at that point to get rid of the typical issues but they are usually confined to comment blocks.

**Comment by d00616** [comment]
@tekka007 hwDebugPrint() is changed. I have added the hwCPUVoltage() functionality.

If nothing more to change, I merge my last commit with 2ae764b2dff36f17987abf93dd3aa2dc2697024d

File: hal/architecture/MyHwNRF5.cpp [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+int8_t hwSleep(uint8_t interrupt1, uint8_t mode1, uint8_t interrupt2,
+               uint8_t mode2, unsigned long ms)
+{
+	// Disable interrupts until going to sleep, otherwise interrupts occurring
+	// between attachInterrupt()
+	// and sleep might cause the MCU to not wakeup from sleep as interrupt has
+	// already be handled!
+	MY_CRITICAL_SECTION {
+		// Fix cppcheck "(style) Variable '__savePriMask' is assigned a value that is never used."
+		(void)__savePriMask;
```
Comment (d00616) [inline hal/architecture/MyHwNRF5.cpp]: I don't have any better idea about getting rid this message. This needs a fix in MyHwNRF5.h and MyHwSAMD.h

Comment (fallberg) [inline hal/architecture/MyHwNRF5.cpp]: (void) casting in these situations is an acceptable solution

**Comment by d00616** [comment]
Now I have merged the review changes into https://github.com/mysensors/MySensors/pull/830/commits/9e9a03f8aeb286e2a21592fff9eb9a84801939a7, so there are two clear commits.

File: drivers/NRF5/Radio_ESB.cpp [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+	if (len > MAX_MESSAGE_LENGTH) {
+		len = MAX_MESSAGE_LENGTH;
+	}
+
+	// copy data to tx_buffer
+	memcpy(&tx_buffer.data[0], buf, len);
+
+	// build metadata
+	tx_buffer.len = len;
+	tx_buffer.noack = noACK;
```
Comment (tekka007) [inline drivers/NRF5/Radio_ESB.cpp]: The RF24 driver sets the NOACK flag either for broadcasted messages or if the noACK parameter is set (used for passive node mode) - I suggest this change:

```tx_buffer.noack = noACK || recipient==BROADCAST_ADDRESS;```

File: drivers/NRF5/Radio_ESB.cpp [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+	tx_buffer.pid++;
+
+	// Calculate number of retries
+	tx_retries = ((recipient == BROADCAST_ADDRESS)?(NRF5_ESB_BC - 1):(NRF5_ESB_ARC - 1));
```
Comment (tekka007) [inline drivers/NRF5/Radio_ESB.cpp]: BC / noACK messages should be sent once.

A general thought: Many users have nRF24L01+ counterfeits with flipped noACK bits - this should be taken into consideration in a mixed (genuine/counterfeit) network, i.e. the SW ESB/RF24 noACK implementation should be adjustable, maybe as preprocessor switch?

File: drivers/NRF5/Radio_ESB.cpp (hunk 2 above)
Comment (d00616) [inline drivers/NRF5/Radio_ESB.cpp]: @tekka007 Changed. If you are fine with this change, I merge the Commit into the main commit.

File: drivers/NRF5/Radio_ESB.cpp (hunk 3 above)
Comment (d00616) [inline drivers/NRF5/Radio_ESB.cpp]: I have added MY_NRF5_ESB_REVERSE_ACK

File: drivers/NRF5/Radio_ESB.h [hunk 4]
```diff
@@ -53,7 +53,7 @@
 #define NRF5_ESB_ARC (15)
 
 // How often broadcast messages are send
-#define NRF5_ESB_BC (4)
+#define NRF5_ESB_BC (1)
```
Comment (tekka007) [inline drivers/NRF5/Radio_ESB.h]: if set to (1) BC msgs won't be sent (due to [this](https://github.com/mysensors/MySensors/pull/830/files/9e9a03f8aeb286e2a21592fff9eb9a84801939a7..60f04e4edd717d8c9d98db04620efc8c2f2883ef#diff-d9fc3fc33e5f695ae5f76dfa3d11fc43L404))

Comment (d00616) [inline drivers/NRF5/Radio_ESB.h]: Thank you. It's fixed.

**Comment by d00616** [comment]
Commits merged.

@tekka007 Thank you for your review.

**Pull request merged**